### DISCUSSION TITLE
Using RF Modules with Raspberry Pi Pico

### DISCUSSION THREAD

**Discussion by Marin-Johnson** [micropython/micropython#10313] [RP2040 / Pico] [answered]
Hello -

I have 2 Raspberry Pi Picos running MicroPython. I am trying to use a 433 MHz [RF transmitter](https://www.digikey.com/en/products/detail/linx-technologies-inc/TXM-433-LR/613950) on one Pico and a 433 MHz [RF receiver ](https://www.digikey.com/en/products/detail/linx-technologies-inc/RXM-433-LR/613947?s=N4IgTCBcDaIEoA0CyBaALAZgygMnEAugL5A)on the other Pico. I am currently using `UART` to transmit and receive data:

    # Receiver

    import os
    import machine
    from time import sleep

    uart = machine.UART(0, 4800)
    print(uart)

    led = machine.Pin(25, machine.Pin.OUT)

    b = None
    msg = ""

    while True:
        if uart.any():
            b = uart.readline()
            try:
                msg = b.decode('utf-8')
                print(str(msg))
            except:
                print("Failed (" + str(type(b)) + "): " + str(b))
                pass
            
        led.toggle()
        sleep(1)

and

    # Transmitter

    import os
    import machine
    from time import sleep

    uart = machine.UART(0, 4800)
    print(uart)

    led = machine.Pin(25, machine.Pin.OUT)

    while True:
        sleep(5)
        led.toggle()
        uart.write('Hello, World!')

But the receiver prints stuff like [this](https://github.com/JustMakeStuff-tech/Demo/blob/main/log.txt) even when the transmitter is not transmitting. (I can't paste it here as it messes with the formatting)

As an experiment, I connected the TX on one Pico to the RX of the other pico, and it was able to send the data successfully. Therefore, I believe it is the transmitter and receiver getting interference from other signals.

**My Question:**
----

Arduino has libraries for packet radio. (see [this](http://www.airspayce.com/mikem/arduino/RadioHead/)) Is there anything similar in MicroPython or for the Raspberry Pi Pico?

Thanks.

**Comment by peterhinch** [comment, marked as answer]
Yes, but not for these modules. See [NRF24l01](https://github.com/micropython/micropython-lib/tree/master/micropython/drivers/radio/nrf24l01) which is a 2.4GHz radio.

Unless a receiver is explicitly designed for use with a UART you can't stick a UART on a receiver and expect it to work. Many receivers output garbage in the absence of a signal.

**Reply by Marin-Johnson** [reply]
Thanks, will use these instead.

**Reply by Marin-Johnson** [reply]
> Unless a receiver is explicitly designed for use with a UART you can't stick a UART on a receiver and expect it to work. Many receivers output garbage in the absence of a signal.

I thought UART should work, as these modules say they support a "Direct serial interface." They also say

> The CMOS-compatible data input on Pin 2 is normally supplied with a serial bit stream from a microprocessor or encoder, but it can also be used with standard UARTs.

So why does UART not work as these modules claim it does?

**Comment by scruss** [comment]
These modules are a step above the horrible cheap ones I've seen in the past. But to echo Peter's answer, a quote from the [https://linxtechnologies.com/wp/wp-content/uploads/txm-fff-lr.pdf](manual):

> … Linx RF modules do not encode or packetize the signal content in any manner

So you have to encode the data yourself, usually a [Manchester encoding](https://en.wikipedia.org/wiki/Manchester_code).

The receiver will typically appear to be only picking up noise. That's because they usually have automatic gain control, and a transmitter has to send a series of training pulses before the data packet to get the receiver's attention. Since there's no error correction, you have to build that into your protocol, too. You also have to be really careful about building the right antenna. Without a library for MicroPython, this will be a lot of work

Peter's NRF recommendation is a good one: they do all the work of building a protocol.

**Reply by Marin-Johnson** [reply]
Got it, thanks.

**Comment by peterhinch** [comment]
[This library](https://github.com/peterhinch/micropython-radio) may be of use for extending the basic functionality of the NRF24l01.
